vespa/src/main/java/ai/marqo/search/HybridSearcher.java

```
@@ -267,21 +288,86 @@ HitGroup rrf(
             }
         }
 
-        // Sort and trim results.
-        logIfVerbose("Combined list (UNSORTED)", verbose);
-        logHitGroup(result, verbose);
+        return result;
+    }
 
-        result.sort();
-        logIfVerbose("Combined list (SORTED)", verbose);
-        logHitGroup(result, verbose);
+    void raiseErrorIfPresent(Result resultLexical, Result resultTensor) {
+        // Raise error if either result list has an error. Make sure error messages are combined
+        String tensorOrLexicalErrors = "";
+        ErrorMessage tensorError = resultTensor.hits().getError();
+        if (tensorError != null) {
+            tensorOrLexicalErrors += "Error in TENSOR search in RRF: " + tensorError;
+        }
 
-        // Only return top hits (max length)
-        Integer finalLength = Math.max(hitsTensor.size(), hitsLexical.size());
-        result.trim(0, finalLength);
-        logIfVerbose("Combined list (TRIMMED)", verbose);
-        logHitGroup(result, verbose);
+        ErrorMessage lexicalError = resultLexical.hits().getError();
+        if (lexicalError != null) {
+            tensorOrLexicalErrors += "Error in LEXICAL search in RRF: " + lexicalError;
+        }
 
-        return result;
+        if (!tensorOrLexicalErrors.isEmpty()) {
+            throw new RuntimeException(tensorOrLexicalErrors);
```

We always throw RuntimeException, which will cause Vespa to return 500 error. Is this expected? Should we distinguish user error from transient server error? 500 server error usually indicates the request is retriable.

---

@vicilliar , you mentioned you also fixed an issue we discussed before. Could you add a comment to highlight the fix? Thanks

---

This is a fair point. We could get more specific perhaps

---

I agree we should have a more sophisticated system to handle the error codes thrown here, but as a different feature.

---

@farshidz Do we have a ticket for the hybrid searcher swallowing errors? I remember one being made but I can't find it on the board

---

Nvm found it: https://s2search.atlassian.net/jira/polaris/projects/MOSD/ideas/view/4863474?selectedIssue=MOSD-225&issueViewLayout=sidebar&issueViewSection=overview&atlOrigin=eyJpIjoiNDdiNzlkMzE4MzUyNGFmZjhkYmM1NDI0NTNjZTNiMjIiLCJwIjoiaiJ9